EIPS/eip-1985.md

```diff
@@ -0,0 +1,118 @@
+by eliminating unrealistic edge cases.
+
+## Specification
+
+If `block.number >= {FORK_BLOCK}`, the following value ranges are introduced.
+They restrict the results (i.e. values pushed to the stack) of the opcodes listed below.
+
+1. *gas*, *gas limit*, *block gas limit*
+   is a range between `0` and `9223372036854775807` (`2**63 - 1`).
+   It affects following the opcodes:
+   - `GASLIMIT` (`0x45`),
+   - `GAS` (`0x5a`).
+   
+2. *block number*, *timestamp*
+   is a range between `0` and `9223372036854775807` (`2**63 - 1`).
+   It affects the following opcodes:
+   - `TIMESTAMP` (`0x42`),
+   - `NUMBER` (`0x43`).
+   
+3. *account address*
+   is a range between `0` and `1461501637330902918203684832716283019655932542975` (`2**160 - 1`).
+   It affects the following opcodes:
+   - `ADDRESS` (`0x30`),
+   - `ORIGIN` (`0x32`),
+   - `CALLER` (`0x33`),
+   - `COINBASE` (`0x41`),
+   - `CREATE` (`0xf0`),
+   - `CREATE2` (`0xf5`). 
+   
+4. *buffer size*
+   is a range between `0` and `4294967295` (`2**32 - 1`).
+   It affects the following opcodes:
+   - `CALLDATASIZE` (`0x36`),
+   - `CODESIZE` (`0x38`),
+   - `EXTCODESIZE` (`0x3b`),
+   - `RETURNDATASIZE` (`0x3d`),
+   - `MSIZE` (`0x59`).
+   
+
+## Rationale
+
+These limits have been:
+- proposed by [EVMC]
+- implemented partially by certain clients, such as [Aleth], [geth], [Parity] and [ethereumjs]
+- allowed by certain test cases in the [Ethereum testing suite]
+- and implicitly also allowed by certain assumptions, such as due to gas limits some of these values cannot grow past a certain limit
+
+Most of the limits proposed in this document have been previously explored and tested in [EVMC].
+
+Using the `2**63 - 1` constant to limit some of the ranges:
+- allows using signed 64-bit integer type to represent it, 
+  what helps programming languages not having unsigned types,
+- makes arithmetic simpler (e.g. checking out-of-gas conditions is simple as `gas_counter < 0`).
+
+### Timestamp
+
+The [Yellow Paper] defines the timestamp in block as "A scalar value equal to the reasonable output of Unix’s time() at this block’s inception".
+IEEE Std 1003.1-2001 (POSIX.1) leaves that definition implementation defined.
+
+### Comparing current implementations
+
+- Timestamp is implemented as a 64-bit value in [geth] and [Parity]
+- Block gas limit is implemented as a 64-bit in [aleth] and [geth]
+
+## Backwards Compatibility
+
+All of these limits are already enforced mostly through the block gas limit. Since the out of range case results in a transaction failure, there should not be a change in behaviour.
+
+## Test Cases
+
+TBA
+
+## Implementation
+
+TBA
+
+## References
+
+[EIP-106](https://github.com/ethereum/EIPs/issues/106) proposed the block gas limit to be limited at `2**63 - 1`.
+
+## TODO
+
+1. The size of addresses are specified in Yellow Paper, 
+e.g. COINBASE is specified as returning H_c which has 20 bytes.
+2. Do the gas limit applies for the gas param in CALL?
```

Comment:
Remove it? What solidity currently provides for calls?

Reply:
I'd say it applies. Then it can be discussed on the ACD calls.